include/mbedtls/config_psa.h

```diff
@@ -675,15 +672,6 @@ extern "C" {
 #define PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY 1
 #endif /* MBEDTLS_ECP_C */
 
-#if defined(MBEDTLS_DHM_C)
```

Comment:
I agree with the other changes (wanting DH in PSA should not enable `DHM_C`) but I'm less sure about that one: in general we want as much as possible supported on the PSA side. So, should we keep this block but just with `BIGNUM_C` as the guard?

Reply:
The description of `MBEDTLS_PSA_CRYPTO_CONFIG` reads:

> When this option is disabled, the PSA API exposes the cryptographic mechanisms that can be implemented on top of the `mbedtls_xxx` API configured with `MBEDTLS_XXX` symbols.

But in fact this looks wrong in the rare cases where a cryptographic mechanism is available through both APIs, but the PSA implementation doesn't use the legacy implementation, so the `PSA_ALG_FOO` “can be implented” without `PSA_FOO_C`. Because “can be implemented” is an internal detail, it isn't a good way to define an interface. Instead, the specification should probably be that PSA exposes cryptographic mechanisms that provide equivalent capabilities.

We have a few precedents which aren't fully consistent:

* The PSA HMAC doesn't rely on the legacy HMAC (provided by `MBEDTLS_MD_C`), only on the availability of the underlying hashes. It is enabled unconditionally.
* The PSA HKDF doesn't rely on the legacy HKDF (provided by `MBEDTLS_HKDF_C`), only on the availability of the underlying hashes. It is enabled if only `MBEDTLS_HKDF_C` is enabled.
* The PSA ECB doesn't rely on the legacy ECB (provided, but only partially, by `MBEDTLS_CIPHER_C`). It is enabled unconditionally if at least one block cipher is enabled.

I think the FFDH situation is closest to HKDF: it would make more sense to provide FFDH through PSA if it's enabled in the legacy interface, regardless of how it's implemented.

HMAC and ECB maybe don't make so much sense, but that's not something we can change without risking breaking existing applications. So I think it'll have to stay that way until the next major release, at which point `MBEDTLS_PSA_CRYPTO_CONFIG` should become always enabled.

(Note that the situation is different when `MBEDTLS_PSA_CRYPTO_CONFIG` is enabled. In that case, legacy modules are only enabled as needed for the implementation, rather than to provide feature parity.)